thewalrus/internal_modes/prepare_cov.py

```diff
@@ -0,0 +1,304 @@
+# Copyright 2021-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Set of functions for forming a covariance matrix over multiple modes, based on overlaps of the squeezer schmidt modes
+"""
+
+from itertools import chain
+
+import numpy as np
+
+from ..decompositions import takagi
+from ..quantum import fidelity
+from ..symplectic import (
+    interferometer,
+    passive_transformation,
+    reduced_state,
+    squeezing,
+)
+
+
+def indexconversion(j, k, njs):
+    r"""
+    Converts the double index (j, k) into a single index = k + \sum_{l=0}^{j-1} njs[l],
+    with njs[0] defined as 0, njs[1] as the number of Schmidt modes in the first spatial mode, etc.
+
+    Args:
+        j (int): spatial mode index
+        k (int): Schmidt mode index
+        njs (array): number of Schmidt modes in each
+
+    Returns:
+        (int): converted index
+    """
+    return k + np.sum(njs[:j])
+
+
+def swap_matrix(M, R):
+    r"""
+    Computes the matrix that swaps the ordering of modes from grouped by orthonormal modes to grouped by spatial modes.
+    The inverse of the swap matrix is its transpose.
+
+    Args:
+        M (int): number of spatial modes
+        R (int): number of orthonormal modes in each spatial mode
+
+    Returns:
+        (array): M*R x M*R swap matrix
+    """
+    P = np.zeros((M * R, M * R), dtype=int)
+    in_modes = range(M * R)
+    out_modes = list(chain.from_iterable(range(i, M * R, M) for i in range(M)))
+    P[out_modes, in_modes] = 1
+    return P
+
+
+def O_matrix(F):
+    r"""The overlap matrix of all of the Schmidt modes in each spatial mode.
+
+    Args:
+        F (list[list/array[array]): List of arrays of the temporal modes of the Schmidt modes in each spatial mode, must be normalized and ordered by spatial mode.
```

Comment:
The data type is missing a closing ']' somewhere. Should it be either list[list[array]] or list[array[array]] (rather that say list[list] or array[array])?

Reply:
There's a missing ], it should be list[list/array[array]]